File: lua/mandel.lua
#!/usr/bin/env lua

local width = 1024
local height = 1024
local header = string.char(
0, -- id length
0, -- color map type
2, -- image type
0, -- first entry index
0,
0, -- color map length
0,
0, -- color map entry size
0, -- x origin
0,
math.mod(height, 256), -- y origin
height / 256,
math.mod(width, 256), -- width
width / 256,
math.mod(height, 256), -- height
height / 256,
24, -- pixel depth
32 -- image descriptor
)

function mandelbrot(ca, cb)
    local za = ca
    local zb = cb

    for i = 1, 128 do
        local zaa = za * za / 1024
        local zbb = zb * zb / 1024

        if zaa + zaa > 4 * 1024 then
            return i
        end

        zb = ((za * zb) / 512) + cb
        za = zaa - zbb + ca
    end

    return 128
end

io.write(header)

for y = 1, height do
    local b = 2 * 1024 - y * 4

    for x = 1, width do
        local a = -2 * 1024 + x * 4
        local v = 256 - 2 * mandelbrot(a, b)

        io.write(string.char(v, v, v))
    end
end

File: php/mandel.php
#!/usr/bin/env php
<?
function putchar($c) {
    printf("%c", $c);
}

function mandelbrot($ca, $cb) {
    global $w;
    global $h;

    $za = $ca;
    $zb = $cb;

    for($i = 0; $i < 128; ++$i) {
        $zaa = $za * $za;
        $zbb = $zb * $zb;

        if($zaa + $zbb > 4)
            return $i;

        $zb = 2 * $za * $zb + $cb;
        $za = $zaa - $zbb + $ca;
    }

    return 127;
}

$w = 1920;
$h = 1200;

putchar(0); // id length
putchar(0); // color map type
putchar(2); // image type
putchar(0); // first entry index
putchar(0);
putchar(0); // color map length
putchar(0);
putchar(0); // color map entry size
putchar(0); // x origin
putchar(0);
putchar($h % 256); // y origin
putchar($h / 256);
putchar($w % 256); // y origin
putchar($w / 256);
putchar($h % 256); // y origin
putchar($h / 256);
putchar(24); // pixel depth
putchar(32); // image descriptor

for($y = 0; $y < $h; ++$y) {
    $b = 2.0 - 4.0 * $y / $h;

    for($x = 0; $x < $w; ++$x) {
        $a = -2.0 + 4.0 * $x / $w;

        $i = 256 - 2 * mandelbrot($a, $b);

        putchar($i);
        putchar($i);
        putchar($i);
    }
}
?>
